lib/DxilDia/DxcPixLiveVariables.cpp

```diff
@@ -0,0 +1,268 @@
+    dxil_debug_info::VariableInfo *VarInfo,
+    unsigned FragmentSizeInBits,
+    unsigned FragmentOffsetInBits
+)
+{
+#ifndef NDEBUG
+  for (unsigned i = 0; i < FragmentSizeInBits; ++i)
+  {
+    const unsigned BitNum = FragmentOffsetInBits + i;
+
+    VarInfo->m_DbgDeclareValidation.resize(
+        std::max<unsigned>(VarInfo->m_DbgDeclareValidation.size(),
+                           BitNum + 1));
+    assert(!VarInfo->m_DbgDeclareValidation[BitNum]);
+    VarInfo->m_DbgDeclareValidation[BitNum] = true;
+  }
+#endif  // !NDEBUG
+}
+
+struct dxil_debug_info::LiveVariables::Impl
+{
+  using VariableInfoMap = std::unordered_map<
+      llvm::DIVariable *,
+      std::unique_ptr<VariableInfo>>;
+
+  using LiveVarsMap =
+      std::unordered_map<llvm::DIScope*, VariableInfoMap>;
+
+  IMalloc *m_pMalloc;
+  DxcPixDxilDebugInfo *m_pDxilDebugInfo;
+  llvm::Module *m_pModule;
+  LiveVarsMap m_LiveVarsDbgDeclare;
+
+  void Init(
+      IMalloc *pMalloc,
+      DxcPixDxilDebugInfo *pDxilDebugInfo,
+      llvm::Module *pModule);
+
+  void Init_DbgDeclare(llvm::DbgDeclareInst *DbgDeclare);
+
+  VariableInfo *AssignValueToOffset(
+      VariableInfoMap *VarInfoMap,
+      llvm::DIVariable *Var,
+      llvm::Value *Address,
+      unsigned FragmentIndex,
+      unsigned FragmentOffsetInBits);
+};
+
+void dxil_debug_info::LiveVariables::Impl::Init(
+    IMalloc *pMalloc,
+    DxcPixDxilDebugInfo *pDxilDebugInfo,
+    llvm::Module *pModule)
+{
+  m_pMalloc = pMalloc;
+  m_pDxilDebugInfo = pDxilDebugInfo;
+  m_pModule = pModule;
+
+  llvm::Function* DbgDeclareFn = llvm::Intrinsic::getDeclaration(
+      m_pModule,
+      llvm::Intrinsic::dbg_declare);
+
+  for (llvm::User* U : DbgDeclareFn->users())
+  {
+    if (auto* DbgDeclare = llvm::dyn_cast<llvm::DbgDeclareInst>(U))
+    {
+      Init_DbgDeclare(DbgDeclare);
+    }
+  }
+}
+
+void dxil_debug_info::LiveVariables::Impl::Init_DbgDeclare(
+    llvm::DbgDeclareInst *DbgDeclare
+)
+{
+  llvm::Value *Address = DbgDeclare->getAddress();
+  auto *Variable = DbgDeclare->getVariable();
+  auto *Expression = DbgDeclare->getExpression();
+
+  if (Address == nullptr || Variable == nullptr || Expression == nullptr)
+  {
+    return;
+  }
+
+  auto* AddressAsAlloca = llvm::dyn_cast<llvm::AllocaInst>(Address);
+  if (AddressAsAlloca == nullptr)
+  {
+      return;
+  }
+
+  auto *S = Variable->getScope();
+  if (S == nullptr)
+  {
+    return;
+  }
+
+  std::unique_ptr<FragmentIterator> Iter = FragmentIterator::Create(
+      DbgDeclare,
+      m_pModule->getDataLayout(),
+      AddressAsAlloca,
+      Expression);
+
+  if (!Iter)
+  {
+    // FragmentIterator creation failure, this skip this var.
+    return;
+  }
+
+  const unsigned FragmentSizeInBits = Iter->FragmentSizeInBits();
+
+  unsigned FragmentIndex;
+  while (Iter->Next(&FragmentIndex))
+  {
+    const unsigned FragmentOffsetInBits =
+        Iter->CurrOffsetInBits();
+
+    VariableInfo* VarInfo = AssignValueToOffset(
+        &m_LiveVarsDbgDeclare[S],
```

Comment:
Could use .at() here so it doesn't accidentally add a new entry in the map?

Reply:
This was intentional -- if there's no entry for S then we want a new one.